### PR TITLE
nixos/coder: add environment.extra and environment.file

### PR DESCRIPTION
**Pull request by gmemstr**
## Description of changes

Adds `environment.file` and `environment.extra` for the Coder module/service. This gives us the ability to pass through secrets and options that can only be configured via environment variables. 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by shyim — APPROVED** [review]
looks good. thank you!

**Comment by kylecarbs** [comment]
@gmemstr is there anything blocking merge of this? I'd love if we could get this in!

**Comment by gmemstr** [comment]
@kylecarbs Nothing blocking from what I'm aware, but I don't have write access to nixpkgs so it needs someone else to merge it in :)

**Comment by kylecarbs** [comment]
@superherointj any chance you could merge?

**Review by superherointj — APPROVED** [review]

File: nixos/modules/services/web-apps/coder.nix
Comment (HopHouse) [inline nixos/modules/services/web-apps/coder.nix:179]: Hello,

Hello, 

I am not sure it is an issue, but I have an error with the module.

I think it should be `cfg.environment.extra` instead of `config.environment.extra`.

Regards.

Comment (superherointj) [reply nixos/modules/services/web-apps/coder.nix:179]: Can you open a PR fixing it?

Comment (HopHouse) [reply nixos/modules/services/web-apps/coder.nix:179]: Thank you for the quick reply. 

I am not 100% confident with both NIX and the PR process. 
But, I will manage to do it during the week-end.

Comment (HopHouse) [reply nixos/modules/services/web-apps/coder.nix:179]: Could not find to do it myself, but @kylecarbs did it : #305993.

**Pull request merged**
